lib/Service/BoardService.php

```diff
@@ -673,15 +675,39 @@ public function clone($id, $userId) {
 
 		$stacks = $this->stackMapper->findAll($id);
 		foreach ($stacks as $stack) {
-			$newStack = new Stack();
-			$newStack->setTitle($stack->getTitle());
-			$newStack->setBoardId($newBoard->getId());
-			$this->stackMapper->insert($newStack);
+			$this->cloneStack($stack, $newBoard->getId());
 		}
 
 		return $newBoard;
 	}
 
+	private function cloneStack($stack, $newBoardId)
+	{
+		$newStack = new Stack();
+		$newStack->setTitle($stack->getTitle());
+		$newStack->setBoardId($newBoardId);
+		$insertedStack = $this->stackMapper->insert($newStack);
+		$cards = $this->cardMapper->findAll($stack->getId());
+
+		foreach ($cards as $card) {
+			$newCard = new Card();
+			$newCard->setTitle($card->getTitle());
+			$newCard->setStackId($insertedStack->getId());
+			$newCard->setType($card->getType());
+			$newCard->setOrder($card->getOrder());
+			$newCard->setOwner($card->getOwner());
+			$newCard->setDescription($card->getDescription());
+			$newCard->setDuedate($card->getDuedate());
+			$newCard = $this->cardMapper->insert($newCard);
```

Comment:
We can isolate this process in a method inside `Card` entity.

```suggestion
			$newCard = new Card();
			$newCard->setFromCard($card);
			$newCard->setStackId($insertedStack->getId());
			$newCard = $this->cardMapper->insert($newCard);
```
or
```suggestion
			$newCard = Card::copyFrom($card);
			$newCard->setStackId($insertedStack->getId());
			$newCard = $this->cardMapper->insert($newCard);
```

Reply:
Yes. But Card is a RelationalEntity. Only accessors and mutators should be defined in it I think. I wanted to add it in CardService but CardService is already using BoardService so there is a dependency issue. What I think is to add a new layer or put it into CardMapper (but mappers are meant to be the ORM layer I think).

Reply:
I see, you are right
Looks like we need to think more about it to create a wise solution.

Reply:
I think we can put the logic in a function just in case we need to isolate later on in a class.